### PR TITLE
Handle evaluation errors in CyIpopt solver

### LINKED ISSUE #2899: Handle evaluation errors in PyNumero
## Summary

Currently, PyNumero's ASL interface handles evaluation errors by asserting the return value of the interface's `eval_g` function [here](https://github.com/Pyomo/pyomo/blob/main/pyomo/contrib/pynumero/asl.py#L426C1-L426C47). To handle this error, e.g. in a solver interface, one would have to catch the error, something like:
```python
try:
    nlp.evaluate_constraints()
except AssertionError:
    # Follow whatever convention the solver uses for eval errors
    return False
```
This is dangerous as it catches any assertion error (e.g. the check for consistent dimensions). I propose we have something like a `PyNumeroEvaluationError` that we raise in this case that can safely be caught and handled by solver interfaces.

Presumably, we could also provide some information about which constraint caused the evaluation error, which would be useful for debugging. Ipopt's AMPL interface does this, although from a quick glance at the ASL evaluation functions it was not obvious to me how this is done.

### Rationale

This is one step towards a CyIpopt interface that handles function evaluation errors in the same way that Ipopt (AMPL interface) does. The other steps are to [handle evaluation errors in CyIpopt](https://github.com/mechmotum/cyipopt/issues/210) and to integrate both of these in our CyIpopt interface.

### PR DESCRIPTION
**Pull request by Robbybp**
## Fixes #2899
(The part of #2899 that was not fixed by #2901)

## Summary/Motivation:
A roadblock to using CyIpopt more often is that our interface did not handle evaluation errors. As of https://github.com/mechmotum/cyipopt/pull/215, CyIpopt handles evaluation errors by allowing callbacks to raise `cyipopt.CyIpoptEvaluationError`, which cause `false` to be returned to Ipopt. When raised in the constraint or objective evaluation, this triggers the step size-reduction routine that many are familiar with through the AMPL interface. This also opens the possibility of displaying more informative debugging information when an evaluation error occurs, although this will likely require tighter integration with ASL and may not be straightforward.

This PR will not pass tests until (a) CyIpopt publishes a release or (b) we test against CyIpopt/main. I'm opening this PR to publicize that this functionality is in the pipeline, but can close it until I expect tests to pass if that's better.

## Changes proposed in this PR:
- Catch `PyNumeroEvaluationError` in `CyIpoptNLP` and raise `CyIpoptEvaluationError`

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by blnicho** [comment]
@Robbybp looks like cyipopt 1.3.0 was released a couple days ago so this PR can now move forward when you have time to come back to it.

**Comment by Robbybp** [comment]
@blnicho Awesome, this is ready for review once tests pass.

File: pyomo/contrib/pynumero/interfaces/cyipopt_interface.py [hunk 1]
```diff
@@ -328,24 +329,44 @@ def scaling_factors(self):
...
     def objective(self, x):
-        self._set_primals_if_necessary(x)
-        return self._nlp.evaluate_objective()
+        try:
+            self._set_primals_if_necessary(x)
+            return self._nlp.evaluate_objective()
+        except PyNumeroEvaluationError:
+            # TODO: halt_on_evaluation_error option. If set, we re-raise the
+            # original exception.
+            raise cyipopt.CyIpoptEvaluationError(
```
Comment (jsiirola) [inline pyomo/contrib/pynumero/interfaces/cyipopt_interface.py:338]: This requires cyipopt 1.3.0.  Should we update the `attempt_import` only allow cyipopt >= 1.3.0, or should we update the importer on line 28 to define a `cyipopt.CyIpoptEvaluationError` exception as a compatibility shim for older versions?  

I would tend to lean toward the latter.

Comment (Robbybp) [reply pyomo/contrib/pynumero/interfaces/cyipopt_interface.py:338]: Good point. I agree with defining a dummy evaluation error for compatibility.

This also raises the question of behavior consistency among different CyIpopt versions. Should the default be to re-raise the original exception, with a `halt_on_evaluation_error` flag that a user can set to `False` if they want to use the new feature?

**Comment by Robbybp** [comment]
Note that this PR changes behavior in our CyIpopt interface somewhat significantly (although the new behavior is closer to the default ASL-Ipopt behavior). Should the default be to halt on these evaluation errors for "backward compatibility" (for all the users who currently rely on CyIpopt solves failing due to evaluation errors...)?

File: pyomo/contrib/pynumero/interfaces/cyipopt_interface.py (hunk 1 above)
Comment (jsiirola) [reply pyomo/contrib/pynumero/interfaces/cyipopt_interface.py:338]: This is in contrib, so backwards compatibility requirements are not as strict. I think I would vote for consistency with the ASL IPOPT interface and have `halt_on_evaluation_error` default to False.

**Comment by Robbybp** [comment]
OSX 3.7 test failure:
```
=================================== FAILURES ===================================
____________________ TestFileUtils.test_find_library_system ____________________

self = <pyomo.common.tests.test_fileutils.TestFileUtils testMethod=test_find_library_system>

    @unittest.skipIf(
        sys.version_info[:2] < (3, 8)
        and (
            platform.mac_ver()[0].startswith('10.16')
            or platform.mac_ver()[0].startswith('12.6')
        ),
        "find_library has known bugs in Big Sur/Monterey for Python<3.8",
    )
    def test_find_library_system(self):
        # Find a system library (before we muck with the PATH)
        _args = {'cwd': False, 'include_PATH': False, 'pathlist': []}
        if FileDownloader.get_sysinfo()[0] == 'windows':
            a = find_library('ntdll', **_args)
            b = find_library('ntdll.dll', **_args)
            c = find_library('foo\\bar\\ntdll.dll', **_args)
        else:
            a = find_library('c', **_args)
            b = find_library('libc.so', **_args)
            c = find_library('foo/bar/libc.so', **_args)
>       self.assertIsNotNone(a)
E       AssertionError: unexpectedly None

pyomo/common/tests/test_fileutils.py:252: AssertionError
------ generated xml file: /Users/runner/work/pyomo/pyomo/TEST-pyomo.xml -------
=========================== short test summary info ============================
FAILED pyomo/common/tests/test_fileutils.py::TestFileUtils::test_find_library_system - AssertionError: unexpectedly None
==== 1 failed, 11169 passed, 5937 skipped, 11 xfailed in 770.13s (0:12:50) =====
Error: Process completed with exit code 1.
```

**Comment by Robbybp** [comment]
Updates:
- Added a `halt_on_evaluation_error` option to `CyIpoptNLP` and `PyomoCyIpoptSolver`. The default behavior depends on whether the CyIpopt version is at least 1.3.0. This is also how we avoid raising `CyIpoptEvaluationError` if it doesn't exit. If `halt_on_evaluation_error`, we re-raise the original exception.
- Tests are run/skipped depending on CyIpopt version, as appropriate

The OSX/3.7 failure appears unrelated, so I believe this is ready for review (assuming no other test failures pop up).

**Comment by jsiirola** [comment]
OSX errors will be "resolved" by #3008

File: pyomo/contrib/pynumero/interfaces/cyipopt_interface.py [hunk 2]
```diff
@@ -328,24 +346,56 @@ def scaling_factors(self):
...
 
     def objective(self, x):
-        self._set_primals_if_necessary(x)
-        return self._nlp.evaluate_objective()
+        try:
+            self._set_primals_if_necessary(x)
+            return self._nlp.evaluate_objective()
+        except PyNumeroEvaluationError:
+            # TODO: halt_on_evaluation_error option. If set, we re-raise the
+            # original exception.
```
Comment (blnicho) [inline pyomo/contrib/pynumero/interfaces/cyipopt_interface.py:354]: Can this TODO be removed? Doesn't this PR add the `halt_on_evaluation_error` option?

**Review by blnicho — APPROVED** [review]
I found one minor thing but otherwise this looks great!

File: pyomo/contrib/pynumero/interfaces/cyipopt_interface.py (hunk 2 above)
Comment (Robbybp) [reply pyomo/contrib/pynumero/interfaces/cyipopt_interface.py]: Yes, that is out of date. Thanks for catching

**Review by michaelbynum — APPROVED** [review]

**Pull request merged**